collect_app/src/test/java/org/odk/collect/android/provider/FormsProviderTest.java

```diff
@@ -0,0 +1,478 @@
+        Uri formUri = addFormsToDirAndDb("external_app_form", "1", "External app form");
+        createFormFileInFormsDir("external_app_form", "1", "External app form renamed");
+
+        String oldCachePath;
+        try (Cursor cursor = contentResolver.query(formUri, null, null, null, null)) {
+            assertThat(cursor.getCount(), is(1));
+
+            cursor.moveToNext();
+            oldCachePath = cursor.getString(cursor.getColumnIndex(JRCACHE_FILE_PATH));
+        }
+
+        ContentValues contentValues = new ContentValues();
+        contentValues.put(JRCACHE_FILE_PATH, "newCache");
+
+        contentResolver.update(formUri, contentValues, null, null);
+
+        assertThat(new File(externalFilesDir + File.separator + ".cache" + File.separator + oldCachePath).exists(), is(false));
+
+        try (Cursor cursor = contentResolver.query(formUri, null, null, null)) {
+            assertThat(cursor.getCount(), is(1));
+
+            cursor.moveToNext();
+            assertThat(cursor.getString(cursor.getColumnIndex(JRCACHE_FILE_PATH)), is("newCache"));
+        }
+    }
+
+    /**
+     * Theory: this is attempting to prevent an edge case where the file path is changed, a new cache path
+     * is generated and then that new cache path is deleted instead of the old one. This was based on
+     * comments in the code that indicated ordering was important.
+     */
+    @Test
+    @Ignore("Not sure we need this")
+    public void update_withFormFilePath_andCachePath_deletesCache() {
+        Uri formUri = addFormsToDirAndDb("external_app_form", "1", "External app form");
+        File newFile = createFormFileInFormsDir("external_app_form", "1", "External app form renamed");
+        String newHash = FileUtils.getMd5Hash(newFile);
+
+        String oldCachePath;
+        try (Cursor cursor = contentResolver.query(formUri, null, null, null, null)) {
+            assertThat(cursor.getCount(), is(1));
+
+            cursor.moveToNext();
+            oldCachePath = cursor.getString(cursor.getColumnIndex(JRCACHE_FILE_PATH));
+        }
+
+        ContentValues contentValues = new ContentValues();
+        contentValues.put(FORM_FILE_PATH, newFile.getAbsolutePath());
+        contentValues.put(JRCACHE_FILE_PATH, "newCache"); // This won't be used as a new cache will be generated
+
+        contentResolver.update(formUri, contentValues, null, null);
+
+        assertThat(new File(externalFilesDir + File.separator + ".cache" + File.separator + oldCachePath).exists(), is(false));
+
+        try (Cursor cursor = contentResolver.query(formUri, null, null, null)) {
+            assertThat(cursor.getCount(), is(1));
+
+            cursor.moveToNext();
+            assertThat(cursor.getString(cursor.getColumnIndex(JRCACHE_FILE_PATH)), is(newHash + ".formdef"));
+        }
+    }
+
+    @Test
+    @Ignore("Not sure we need this")
+    public void update_withSelection_andFormFilePath_deletesFiles_andUpdatesHashAndPaths() {
```

Comment:
It does feel important that an update to form file path would update the md5 hash.

As I've alluded to above, an important question is whether there's any value to passing in a form path now that only paths relative to Collect's storage can be used. Could apps write relative paths that reach their own dirs?

Reply:
None of these deal with "fast external itemset" data, right?

Reply:
> It does feel important that an update to form file path would update the md5 hash

Yeah agreed. There's actually a test for that in the `FormsRepositoryTest` so that is handled.

> As I've alluded to above, an important question is whether there's any value to passing in a form path now that only paths relative to Collect's storage can be used. Could apps write relative paths that reach their own dirs?

Maybe? I guess because of how `StoragePathProvider#getRelativePath` works if an external app gives us paths in their own dir they would be stored as absolute (as they don't start with our forms dir). I think Collect would break if it tries to access these forms though as it would prepend our forms dir path. If we want to support that kind of behaviour it would be pretty easy to write a test for it (just insert a form that lives somewhere else on the device and check Collect can open it for instance.

> None of these deal with "fast external itemset" data, right?

Nope that only got deleted as part of `delete`. Remember these are tests for the existing functionality so I haven't tried to alter or fix any behaviour.